### PR TITLE
Make resource constructor parameters writables

### PR DESCRIPTION
**Pull request by Nek-**
This is motivated because other tools are already constructor friendly (ie Symfony serializer and Doctrine).

Also using constructors must be recommended and not supporting them is a serious feature missing.

| Q             | A
| ------------- | ---
| Bug fix?      | somehow
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | https://github.com/api-platform/core/issues/1747
| License       | MIT
| Doc PR        | N/A

### CODE REVIEW

File: src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+        $ref = new \ReflectionClass($resourceClass);
+        if (!$ref->isInstantiable()) {
+            return $propertyMetadata;
+        }
+
+        $constructorParameters = $ref->getConstructor()->getParameters();
+        foreach ($constructorParameters as $constructorParameter) {
+            if ($constructorParameter->getName() === $property) {
+                return $propertyMetadata->withWritable(true);
```
Comment (soyuka) [inline src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: Okay but this doesn't mean that the PropertyAccess component will be able to write to these properties right (denormalization process)?

Comment (dunglas) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: The PropertyAccessor component directly no, but the Serializer component knows how to use constructors, and constructor support has been dramatically improved in 4.1.

Comment (soyuka) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: Cool then :).

File: src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        } else {
+            try {
+                $propertyMetadata = $this->decorated->create($resourceClass, $property, $options);
+            } catch (PropertyNotFoundException $propertyNotFoundException) {
+                $propertyMetadata = new PropertyMetadata();
+            }
+        }
+
+        // Constructor arguments are obviously accessible only on post operation, put will result in an error.
+        if (!isset($options['collection_operation_name']) || 'post' !== $options['collection_operation_name']) {
```
Comment (soyuka) [inline src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: What if we have custom collection operations that don't have the `post` name?

File: src/Bridge/Symfony/Bundle/Resources/config/metadata/metadata.xml [hunk 3]
```diff
@@ -69,6 +69,10 @@
             <argument type="service" id="api_platform.metadata.property.metadata_factory.serializer.inner" />
         </service>
 
+        <service id="api_platform.metadata.property.metadata_factory.constructor" class="ApiPlatform\Core\Metadata\Property\Factory\ConstructorPropertyMetadataFactory" decorates="api_platform.metadata.property.metadata_factory" decoration-priority="20" public="false">
```
Comment (soyuka) [inline src/Bridge/Symfony/Bundle/Resources/config/metadata/metadata.xml]: Could we use 30 as a priority as well?

File: src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php (hunk 2 above)
Comment (Nek-) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: I don't understand the use case, can you be more specific? It doesn't make sense to allow constructor in another request type than post.

File: src/Bridge/Symfony/Bundle/Resources/config/metadata/metadata.xml (hunk 3 above)
Comment (Nek-) [reply src/Bridge/Symfony/Bundle/Resources/config/metadata/metadata.xml]: Actually 35 makes even more sense (as the serializer `SerializerPropertyMetadataFactory` will potentially modify this accessibility) !

File: src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php (hunk 2 above)
Comment (meyerbaptiste) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: He meant this:
```
collectionOperations={
    "foo" => {"method"="POST"}
}
```

Comment (Nek-) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: I guess the best way to fix this is to add the method to the context, what do you think about?

Comment (meyerbaptiste) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: Huum, `ResourceMetadata` should do the job. For example:
```php
$this->resourceMetadataFactory->create($resourceClass)->getCollectionOperationAttribute($options['collection_operation_name'], 'method')
```

Comment (Nek-) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: Fixed.

**Comment by Nek-** [comment]
In what cases is this possible?

```php
$method = $this->resourceMetadataFactory
    ->create($resourceClass)
    ->getCollectionOperationAttribute($options['collection_operation_name'], 'method')
;
// $method === null
```

**Comment by meyerbaptiste** [comment]
There is a better way to do that with the `OperationMethodResolver` class:
```php
$this->operationMethodResolver->getCollectionOperationMethod($resourceClass, $options['collection_operation_name']);
```
Sorry, I missed this class...

**Comment by Nek-** [comment]
Indeed. Fixed!

**Comment by dunglas** [comment]
I though it would be harder! Good job!

Would you mind an integration test to be sure we support this new feature everywhere?

File: src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        }
+
+        if (!isset($options['collection_operation_name'])) {
+            return $propertyMetadata;
+        }
+
+        // Constructor arguments are obviously accessible only on post operation, put will result in an error.
+        try {
+            $method = $this->operationMethodResolver->getCollectionOperationMethod($resourceClass, $options['collection_operation_name']);
+        } catch (RuntimeException $e) {
```
Comment (soyuka) [inline src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: when does this throw a RuntimeException?

File: tests/Metadata/Property/Factory/ConstructorPropertyMetadataFactoryTest.php [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    private function getMethodResolverMock($method = 'post')
+    {
+        $resolver = $this->prophesize(OperationMethodResolverInterface::class);
+        $resolver->getCollectionOperationMethod(Argument::any(), Argument::any())->willReturn($method);
+
+        return $resolver->reveal();
+    }
+}
+
+class DummyObjectWithConstructor
```
Comment (soyuka) [inline tests/Metadata/Property/Factory/ConstructorPropertyMetadataFactoryTest.php]: Wasn't best practice to put these in another file?

File: src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php (hunk 4 above)
Comment (Nek-) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: It's not a problem if the method is not available.

Comment (Nek-) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: I also asked :) : https://github.com/api-platform/core/pull/1749#issuecomment-371136092

Comment (meyerbaptiste) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: Here: https://github.com/api-platform/core/blob/4fa2e4a7665dabdd29487fa907708ae5ef907c7e/src/Bridge/Symfony/Routing/OperationMethodResolver.php#L96

File: tests/Metadata/Property/Factory/ConstructorPropertyMetadataFactoryTest.php (hunk 5 above)
Comment (Nek-) [reply tests/Metadata/Property/Factory/ConstructorPropertyMetadataFactoryTest.php]: Well, that depends on your POV. IMO re-using dummy classes in many test is not a good practice because it means your test is dependent on something _outside_ (see [FIRST](https://stackoverflow.com/questions/18024785/tdd-first-principle)). I do it only in case I need doctrine entities to avoid configuration duplication.

TL;DR: I don't think it's a big deal.

Comment (soyuka) [reply tests/Metadata/Property/Factory/ConstructorPropertyMetadataFactoryTest.php]: I agree but it's how we choose to do in this project and we have to follow the project rules :).

Comment (soyuka) [reply tests/Metadata/Property/Factory/ConstructorPropertyMetadataFactoryTest.php]: I know for a fact that I did the same as you and I was told to move my classes out of the test file that's all :p.

File: src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+     * @var PropertyMetadataFactoryInterface
+     */
+    private $decorated;
+
+    /**
+     * @var OperationMethodResolverInterface
+     */
+    private $operationMethodResolver;
+
+    public function __construct(OperationMethodResolverInterface $operationMethodResolver, PropertyMetadataFactoryInterface $decorated = null)
```
Comment (Taluu) [inline src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: Why would the decorated object be `null` ?

Comment (soyuka) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: so that it can be used without decoration

File: tests/Metadata/Property/Factory/ConstructorPropertyMetadataFactoryTest.php (hunk 5 above)
Comment (Nek-) [reply tests/Metadata/Property/Factory/ConstructorPropertyMetadataFactoryTest.php]: So I guess it's a rule :) .

File: src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php (hunk 6 above)
Comment (Taluu) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: IMO it's either you decorate something or you don't, Not both at the same time (S from Solid, anyone ?).

Having a basic decorator that just returns a `new PropertyMetadata()` would be simpler.

Comment (dunglas) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: @Taluu loaders can be chained in any order (and sometimes you have to change the order, to change the priority). The first registered loader will not be a decorator, but you cannot guess which one will be the first when writing the code (it depends of the config).

Comment (Taluu) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: Then it means it is the wrong design pattern for these. Shouldn't a strategy be better for that ? of course, nothing prevents some elements in this pattern to be decorators...

Comment (Taluu) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: Or a chain of responsability would also do it

Comment (dunglas) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: It was tried, and they introduces a lof of unneeded complexity. They have been a lot of iteration on this part, I don't want to change things again. It's mostly internal, and it works well as is.

Comment (dunglas) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: +using a decorator makes it very easy to configure with the Symfony DI component since it supports service decoration and priority. Using a different pattern such as chain (as it was in API Platform 1) will require to add a lot of specific code to have the same flexibility.

File: src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php (hunk 4 above)
Comment (Nek-) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: It was about how the method can be null (well, ok, the route is null, but it's not the answer).

File: src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        }
+
+        // Constructor arguments are obviously accessible only on post operation, put will result in an error.
+        try {
+            $method = $this->operationMethodResolver->getCollectionOperationMethod($resourceClass, $options['collection_operation_name']);
+        } catch (RuntimeException $e) {
+            return $propertyMetadata;
+        }
+
+        if ('post' !== strtolower($method)) {
```
Comment (meyerbaptiste) [inline src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: With #1752, `'POST' !== $method` is enough. You just have to rebase.

File: src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+/**
+ * Set properties available in the constructor as writable.
+ *
+ * @author Maxime Veber <maxime.veber@nekland.fr>
+ */
+final class ConstructorPropertyMetadataFactory implements PropertyMetadataFactoryInterface
+{
+    /**
+     * @var PropertyMetadataFactoryInterface
```
Comment (dunglas) [inline src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: Useless (documented by the typehint), to remove.

File: src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+     */
+    private $decorated;
+
+    /**
+     * @var OperationMethodResolverInterface
```
Comment (dunglas) [inline src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: Useless (documented by the typehint), to remove.

File: tests/Metadata/Property/Factory/ConstructorPropertyMetadataFactoryTest.php (hunk 5 above)
Comment (dunglas) [reply tests/Metadata/Property/Factory/ConstructorPropertyMetadataFactoryTest.php]: It's a rule, because it's what recommended is by the PSRs. Anyway, we try to enforce **not** reusing dummy classes, because of first and - especially - because it's a pain to maintain such tests.

Both can be done together: a test = a dummy class in a separate file. It's what we should enforce.

File: src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php (hunk 8 above)
Comment (Taluu) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: Not all editors / etc uses the constructor typehint for properties, so this looks adequate ?

Comment (dunglas) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: We do that everywhere in the project, and in Symfony as well (we follow Symfony's CS). It's less maintenance and PHPStorm supports this so... 😀

Comment (Taluu) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: yeah, that's why it was more a question, but then I opened a ticket on phpactor. :D

File: tests/Metadata/Property/Factory/ConstructorPropertyMetadataFactoryTest.php (hunk 5 above)
Comment (Taluu) [reply tests/Metadata/Property/Factory/ConstructorPropertyMetadataFactoryTest.php]: Couldn't an anonymous class be used ?

File: src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+            return $propertyMetadata;
+        }
+
+        $ref = new \ReflectionClass($resourceClass);
+        if (!$ref->isInstantiable() || !$constructor = $ref->getConstructor()) {
+            return $propertyMetadata;
+        }
+
+        foreach ($constructor->getParameters() as $constructorParameter) {
+            if ($constructorParameter->getName() === $property) {
```
Comment (teohhanhui) [inline src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: I think we could just do `$constructorParameter->name`? It seems to be the same.

File: tests/Metadata/Property/Factory/ConstructorPropertyMetadataFactoryTest.php (hunk 5 above)
Comment (teohhanhui) [reply tests/Metadata/Property/Factory/ConstructorPropertyMetadataFactoryTest.php]: Putting them outside doesn't mean reusing them. We should not have more than 1 class in a file, as that wouldn't work for PSR-4 autoloading.

**Comment by Nek-** [comment]
Fixes done. Rebased. Behat tests added. 😄

File: src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+            return $propertyMetadata;
+        }
+
+        $ref = new \ReflectionClass($resourceClass);
+        if (!$ref->isInstantiable() || !$constructor = $ref->getConstructor()) {
+            return $propertyMetadata;
+        }
+
+        foreach ($constructor->getParameters() as $constructorParameter) {
+            if ($constructorParameter->name === $property) {
```
Comment (dunglas) [inline src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: We must also check that the current value of `writable` is null (to avoid overwriting a manual setting tru `@ApiProperty`).

Comment (Nek-) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: Hum. I think modifying the order would be better because the property accessor will set this value to "false".

Actually, this issue is also present in the `ExtractorPropertyMetadataFactory`: if the property accessor decided to set the property to "not accessible" (aka `false`), no manual configuration can reverse the value because it fills it only if the current value is "null".

What do you think of that?

Comment (teohhanhui) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: The user's manual configuration must always be able to override anything. And yes, I think this decorator should run after the PropertyInfo decorator.

Comment (Nek-) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: Ok, so this answer means that there's an issue inside the `ExtractorPropertyMetadataFactory`.

It also means as I understand that I can indeed override the `writable` value (so I don't modify this line).

**Comment by Nek-** [comment]
I see that the problem I notice does not generate eruption of volunteers to check if there's actually something wrong :)... So here is a schema of what's happen (as far as I understand). The problem is more obvious if you look at it. (you may also see some other issues in it)

apiplatformpropertyfactory

**Comment by Nek-** [comment]
I fixed my first commit and added a new commit for the new things pointed out.

File: src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+namespace ApiPlatform\Core\Metadata\Property\Factory;
+
+use ApiPlatform\Core\Api\OperationMethodResolverInterface;
+use ApiPlatform\Core\Exception\PropertyNotFoundException;
+use ApiPlatform\Core\Exception\RuntimeException;
+use ApiPlatform\Core\Metadata\Property\PropertyMetadata;
+
+/**
+ * Set properties available in the constructor as writable.
```
Comment (dunglas) [inline src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: Sets

File: src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php [hunk 13]
```diff
@@ -0,0 +1,80 @@
+<?php
```
Comment (dunglas) [inline src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: Do we really need this class? Can't we use https://github.com/symfony/symfony/pull/25605 directly?
https://github.com/api-platform/core/blob/master/src/Bridge/Symfony/PropertyInfo/Metadata/Property/PropertyInfoPropertyMetadataFactory.php#L67 will do the job

Comment (Nek-) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: You're suggesting me to add things non-related to the property metadata factory inside of it. I don't think it's a great idea (it's about single responsibility).

It also enforces this feature to be inside the Symfony bridge (this is not a huge issue for me but all this work of decoupling for finally add things that may be decoupled easily from Symfony sounds sad to me).

File: src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php (hunk 12 above)
Comment (Nek-) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: fixed

File: src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php (hunk 13 above)
Comment (dunglas) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: I’m not sure to understand your last comment. To clarify: if we drop this class, and bump PropertyInfo to 4.1 it will work without further modification right?

**Comment by bendavies** [comment]
I'd like to use this. Do we need help finishing it off?

File: src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php (hunk 13 above)
Comment (soyuka) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: Yes it should, @Nek- we just need to enforce `property-info` to `4.1` and it should take care of handling the constructor things for us.

Comment (dunglas) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: We only have to enforce it in our dev deps, with an older version, constructor support will not work but that's all.

Comment (Nek-) [reply src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php]: https://github.com/symfony/symfony/pull/25605 is about getting the type of a property via the constructor. I don't understand why you are speaking about it like it could replace the code of this class. Can you tell me more ?

**Comment by soyuka** [comment]
I think that this is working already with `property-info@4.1`.

**Comment by Nek-** [comment]
Just tested to be really sure.

```php
class Foo
{
    private $bar;

    public function __construct($bar)
    {
        $this->bar = $bar;
    }
}

$reflectionExtractor = new \Symfony\Component\PropertyInfo\Extractor\ReflectionExtractor();

$propertyInfo = new \Symfony\Component\PropertyInfo\PropertyInfoExtractor([$reflectionExtractor], [$reflectionExtractor],[], [$reflectionExtractor]);

$properties = $propertyInfo->getProperties(Foo::class);
$writable = $propertyInfo->isWritable(Foo::class, 'bar');

var_dump($properties);
var_dump($writable);
```

And here is the output:
```
NULL
bool(false)
```

(With `symfony/property-info:@dev`)

**Comment by bendavies** [comment]
Can the feature be made to work with the creation of relations of resources?

**Comment by Nek-** [comment]
This feature is not "limited". It works for what it is designed for.

But please tell me what do you think (example of usage) because I don't understand the feature you speak about.

**Comment by bendavies** [comment]
hey @Nek-, sorry, I didn't mean to offend.
What i meant was that API-Platform supports creating relations like this:
https://github.com/api-platform/core/blob/342b92a3d15cf6c9fa610872813f92d45fc36179/features/main/relation.feature#L361

Or anything in that feature really.

**Comment by dunglas** [comment]
Maybe could we copy https://github.com/symfony/symfony/pull/26997 in API Platform for now and mark those classes and interfaces as internal? We'll drop them when it will be merged in Symfony. WDYT?

**Comment by Nek-** [comment]
I good with that, I'll be working on a refactoring ASAP.

**Comment by seb2411** [comment]
Hello, 

Can I extract the fix related to the Metadata Factory decoration priority and create a separate PR for this ?

Thx

**Comment by bendavies** [comment]
what is the bug the priority change is fixing? please see https://github.com/api-platform/core/issues/1980

**Comment by Nek-** [comment]
@dunglas this PR should be ready. What do you think?

File: src/Bridge/Symfony/PropertyInfo/Metadata/Property/PropertyInfoPropertyMetadataFactory.php [hunk 14]
```diff
@@ -68,6 +68,23 @@ public function create(string $resourceClass, string $name, array $options = [])
             $propertyMetadata = $propertyMetadata->withWritable($writable);
         }
 
+        if (method_exists($this->propertyInfo, 'isInitializable')) {
+            if (true !== $propertyMetadata->isInitializable() && null !== $initializable = $this->propertyInfo->isInitializable($resourceClass, $name, $options)) {
```
Comment (meyerbaptiste) [inline src/Bridge/Symfony/PropertyInfo/Metadata/Property/PropertyInfoPropertyMetadataFactory.php]: IMO it should be `null !== $propertyMetadata->isInitializable()`.

**Comment by Nek-** [comment]
I think issues are non-related.

**Comment by Nek-** [comment]
I removed the part that was suppose to fix order in metadata factories: it was failing. This will still be an issue but this PR should now pass tests.

**Comment by soyuka** [comment]
> I removed the part that was suppose to fix order in metadata factories: it was failing. This will still be an issue but this PR should now pass tests.

I guess we can track this in https://github.com/api-platform/core/issues/1980 

You have a small indentation issue in `DummyEntityWithConstructor`, just run php-cs-fixer and we're good to go on this :).

**Comment by Nek-** [comment]
Fixed 👍

File: src/Serializer/AbstractItemNormalizer.php [hunk 15]
```diff
@@ -142,8 +142,13 @@ protected function getAllowedAttributes($classOrObject, array $context, $attribu
 
             if (
                 $this->isAllowedAttribute($classOrObject, $propertyName, null, $context) &&
-                ((isset($context['api_normalize']) && $propertyMetadata->isReadable()) ||
-                (isset($context['api_denormalize']) && $propertyMetadata->isWritable()))
+                (
+                    (isset($context['api_normalize']) && $propertyMetadata->isReadable()) ||
```
Comment (teohhanhui) [inline src/Serializer/AbstractItemNormalizer.php]: Unnecessary brackets...

File: src/Serializer/AbstractItemNormalizer.php [hunk 16]
```diff
@@ -142,8 +142,13 @@ protected function getAllowedAttributes($classOrObject, array $context, $attribu
...
+                    (
```
Comment (teohhanhui) [inline src/Serializer/AbstractItemNormalizer.php]: Unnecessary brackets...

File: src/Serializer/AbstractItemNormalizer.php [hunk 17]
```diff
@@ -142,8 +142,13 @@ protected function getAllowedAttributes($classOrObject, array $context, $attribu
 
             if (
```
Comment (teohhanhui) [inline src/Serializer/AbstractItemNormalizer.php]:
```php
if (
    $this->isAllowedAttribute($classOrObject, $propertyName, null, $context) &&
    (
        isset($context['api_normalize']) && $propertyMetadata->isReadable() ||
        isset($context['api_denormalize']) && ($propertyMetadata->isWritable() || !is_object($classOrObject) && $propertyMetadata->isInitializable())
    )
) {
    $allowedAttributes[] = $propertyName;
}
```

File: tests/Fixtures/TestBundle/Entity/DummyEntityWithConstructor.php [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+     * @var int The id
+     *
+     * @ORM\Column(type="integer")
+     * @ORM\Id
+     * @ORM\GeneratedValue(strategy="AUTO")
+     */
+    private $id;
+
+    /**
+     * @var string The tour name
```
Comment (teohhanhui) [inline tests/Fixtures/TestBundle/Entity/DummyEntityWithConstructor.php]: Remove PHPDoc descriptions that don't make sense... :smile:

File: tests/Fixtures/TestBundle/Entity/DummyEntityWithConstructor.php [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+     *
+     * @ORM\Column
+     */
+    private $foo;
+
+    /**
+     * @var string The tour name
```
Comment (teohhanhui) [inline tests/Fixtures/TestBundle/Entity/DummyEntityWithConstructor.php]: Remove PHPDoc descriptions that don't make sense... :smile:

File: tests/Fixtures/TestBundle/Entity/DummyEntityWithConstructor.php [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+     *
+     * @ORM\Column
+     */
+    private $bar;
+
+    /**
+     * @var string The tour name
```
Comment (teohhanhui) [inline tests/Fixtures/TestBundle/Entity/DummyEntityWithConstructor.php]: Remove PHPDoc descriptions that don't make sense... :smile:

File: src/Serializer/AbstractItemNormalizer.php (hunk 16 above)
Comment (Nek-) [reply src/Serializer/AbstractItemNormalizer.php]: Looks more clear with too me.

File: src/Serializer/AbstractItemNormalizer.php (hunk 17 above)
Comment (Nek-) [reply src/Serializer/AbstractItemNormalizer.php]: Long lines should be avoided, isn't it?

**Comment by komik966** [comment]
I'm closing my PR, because it's almost done here. Nice work. If you decide to add tests for relations feel free to copy it from my PR.

File: src/Serializer/AbstractItemNormalizer.php (hunk 17 above)
Comment (teohhanhui) [reply src/Serializer/AbstractItemNormalizer.php]: Split across more lines != easier to read :smile:

Comment (teohhanhui) [reply src/Serializer/AbstractItemNormalizer.php]: I think many nested levels of brackets, especially with uneven indentation levels, is much worse than long lines.

**Comment by Nek-** [comment]
@komik966 as it is your work it's probably better that you do the PR for the test on relations. (which is nice)

**Comment by bendavies** [comment]
this can't be merged without tests for relations. let's get them in somehow! @komik966 can PR to @Nek-'s branch?

File: src/Serializer/AbstractItemNormalizer.php (hunk 17 above)
Comment (Nek-) [reply src/Serializer/AbstractItemNormalizer.php]: ✅

File: src/Bridge/Symfony/PropertyInfo/Metadata/Property/PropertyInfoPropertyMetadataFactory.php [hunk 21]
```diff
@@ -68,6 +68,23 @@ public function create(string $resourceClass, string $name, array $options = [])
             $propertyMetadata = $propertyMetadata->withWritable($writable);
         }
 
+        if (method_exists($this->propertyInfo, 'isInitializable')) {
+            if (null === $propertyMetadata->isInitializable() && null !== $initializable = $this->propertyInfo->isInitializable($resourceClass, $name, $options)) {
+                $propertyMetadata = $propertyMetadata->withInitializable($initializable);
+            }
+        } else {
+            // BC layer for Symfony < 4.1
```
Comment (dunglas) [inline src/Bridge/Symfony/PropertyInfo/Metadata/Property/PropertyInfoPropertyMetadataFactory.php]: Symfony 4.2 actually.

File: src/Metadata/Property/PropertyMetadata.php [hunk 22]
```diff
@@ -34,8 +34,9 @@
     private $childInherited;
     private $attributes;
     private $subresource;
+    private $initializable;
 
-    public function __construct(Type $type = null, string $description = null, bool $readable = null, bool $writable = null, bool $readableLink = null, bool $writableLink = null, bool $required = null, bool $identifier = null, string $iri = null, $childInherited = null, array $attributes = null, SubresourceMetadata $subresource = null)
+    public function __construct(Type $type = null, string $description = null, bool $readable = null, bool $writable = null, bool $readableLink = null, bool $writableLink = null, bool $required = null, bool $identifier = null, string $iri = null, $childInherited = null, array $attributes = null, SubresourceMetadata $subresource = null, $initializable = null)
```
Comment (dunglas) [inline src/Metadata/Property/PropertyMetadata.php]: `bool $initializable = null`

File: src/Bridge/Symfony/PropertyInfo/Metadata/Property/PropertyInfoPropertyMetadataFactory.php (hunk 21 above)
Comment (Nek-) [reply src/Bridge/Symfony/PropertyInfo/Metadata/Property/PropertyInfoPropertyMetadataFactory.php]: Arf, yes.

**Comment by soyuka** [comment]
@Nek- @komik966 if you want me to take care of adding the tests written in #1843 in this patch let me know

**Comment by Nek-** [comment]
@soyuka I prefer to not add those tests in this MR if you don't mind.

File: src/Bridge/Symfony/PropertyInfo/Metadata/Property/PropertyInfoPropertyMetadataFactory.php [hunk 23]
```diff
@@ -68,6 +68,23 @@ public function create(string $resourceClass, string $name, array $options = [])
             $propertyMetadata = $propertyMetadata->withWritable($writable);
         }
 
+        if (method_exists($this->propertyInfo, 'isInitializable')) {
+            if (null === $propertyMetadata->isInitializable() && null !== $initializable = $this->propertyInfo->isInitializable($resourceClass, $name, $options)) {
+                $propertyMetadata = $propertyMetadata->withInitializable($initializable);
+            }
+        } else {
+            // BC layer for Symfony < 4.2
+            // To be removed in EOF of Symfony 3.4
```
Comment (teohhanhui) [inline src/Bridge/Symfony/PropertyInfo/Metadata/Property/PropertyInfoPropertyMetadataFactory.php]: I don't understand this comment. If it's a BC layer for Symfony < 4.2, shouldn't it only be removed when we drop support for Symfony < 4.2?

Comment (dunglas) [reply src/Bridge/Symfony/PropertyInfo/Metadata/Property/PropertyInfoPropertyMetadataFactory.php]: Yes we can just drop the last line

**Comment by komik966** [comment]
added tests in #1991 
without modifications in `ApiPlatform\Core\Serializer\AbstractItemNormalizer` relations will not work

**Comment by teohhanhui** [comment]
The tests should be in this PR.

**Comment by komik966** [comment]
to add tests here @Nek- can rebase from my branch

File: src/Bridge/Symfony/PropertyInfo/Metadata/Property/PropertyInfoPropertyMetadataFactory.php (hunk 23 above)
Comment (Nek-) [reply src/Bridge/Symfony/PropertyInfo/Metadata/Property/PropertyInfoPropertyMetadataFactory.php]: Yes my bad!

**Comment by Nek-** [comment]
Okok guys, I was more in favor of merging PRs separately. But well, if you disagree then please review tests in https://github.com/api-platform/core/pull/1991 first because they are not ok to me (and I assume also not for you).

**Comment by dunglas** [comment]
Thank you very much for this @Nek-! It's very appreciated.

**Comment by bendavies** [comment]
great, can't wait to try this.

@komik966 can you work on the tests now?

**Comment by komik966** [comment]
In June I've limited time, so work may be little slow. But I think we are very close to fully working relations. Partially it was implemented in my PR (without handling circular references).

**Comment by teohhanhui** [comment]
This is great work, but I think it's unfortunate that we've merged this without tests.

**Comment by Nek-** [comment]
@teohhanhui it is actually tested.

**Comment by bendavies** [comment]
Can we just clarify what needs to be done here so we don't leave work to be done, as i'm a little confused.

From reading above @Nek-, you didn't want @komik966's tests in your branch, but now you say it is actually tested?

Thanks!

**Comment by Nek-** [comment]
@bendavies this is a test: https://github.com/api-platform/core/pull/1749/files#diff-b63a56b1101fe1091a9cf488e2b0b383R1

The branch of komik966 is nice because it adds many other tests, but it was not mergeable as it to me (and not only to me as I understand).

What's the problem? I don't understand.

**Comment by soyuka** [comment]
@komik966's tests are covering relations from what I can see and there might be an issue there. Also, because of the red coverage, I think that this could be more tested then it is already :).
